server/src/test/java/org/opensearch/search/backpressure/SearchBackpressureServiceTests.java

```diff
@@ -335,6 +393,181 @@ public void testSearchShardTaskInFlightCancellation() {
         assertEquals(expectedStats, actualStats);
     }
 
+    public void testNonCancellationOfHeapBasedTasksWhenHeapNotInDuress() {
+        TaskManager mockTaskManager = spy(taskManager);
+        TaskResourceTrackingService mockTaskResourceTrackingService = mock(TaskResourceTrackingService.class);
+        AtomicLong mockTime = new AtomicLong(0);
+        LongSupplier mockTimeNanosSupplier = mockTime::get;
+
+        EnumMap<ResourceType, NodeDuressTracker> duressTrackers = new EnumMap<>(ResourceType.class) {
+            {
+                put(JVM, new NodeDuressTracker(() -> false, () -> 3));
+                put(CPU, new NodeDuressTracker(() -> true, () -> 3));
+            }
+        };
+
+        NodeDuressTrackers nodeDuressTrackers = new NodeDuressTrackers(duressTrackers);
+
+        // Creating heap and cpu usage trackers where heap tracker will always evaluate with reasons to cancel the
+        // tasks but heap based cancellation should not happen because heap is not in duress
+        TaskResourceUsageTracker heapUsageTracker = getMockedTaskResourceUsageTracker(
+            TaskResourceUsageTrackerType.HEAP_USAGE_TRACKER,
+            (task) -> Optional.of(new TaskCancellation.Reason("mem exceeded", 10))
+        );
+        TaskResourceUsageTracker cpuUsageTracker = getMockedTaskResourceUsageTracker(
+            TaskResourceUsageTrackerType.CPU_USAGE_TRACKER,
+            (task) -> {
+                if (task.getTotalResourceStats().getCpuTimeInNanos() < 400) {
+                    return Optional.empty();
+                }
+                return Optional.of(new TaskCancellation.Reason("cpu time limit exceeded", 5));
+            }
+        );
+
+        TaskResourceUsageTrackers taskResourceUsageTrackers = new TaskResourceUsageTrackers();
+        taskResourceUsageTrackers.addTracker(cpuUsageTracker, TaskResourceUsageTrackerType.CPU_USAGE_TRACKER);
+        taskResourceUsageTrackers.addTracker(heapUsageTracker, TaskResourceUsageTrackerType.HEAP_USAGE_TRACKER);
+
+        // Mocking 'settings' with predictable rate limiting thresholds.
+        SearchBackpressureSettings settings = getBackpressureSettings("enforced", 0.1, 0.003, 10.0);
+
+        SearchBackpressureService service = new SearchBackpressureService(
+            settings,
+            mockTaskResourceTrackingService,
+            threadPool,
+            mockTimeNanosSupplier,
+            nodeDuressTrackers,
+            taskResourceUsageTrackers,
+            new TaskResourceUsageTrackers(),
+            mockTaskManager
+        );
+
+        service.doRun();
+        service.doRun();
```

Comment:
duplicate ?

Reply:
Also don't see any verification that indeed no tasks were treated for cancellation with above thresholds.

Reply:
The cancellation logic will trigger when node is in duress due to any of the system resource. The duress is decided when the duress tracker for a resource returns `true` consecutively for 3 times in a row.

Hence the third invocation to `service.doRun()` will trigger the cancellation logic.

Reply:
the check is little further in this method which is after the cancellation is triggered at line 471.
```SearchBackpressureStats expectedStats = new SearchBackpressureStats(
            new SearchTaskStats(
                5,
                1,
                0,
                Map.of(
                    TaskResourceUsageTrackerType.CPU_USAGE_TRACKER,
                    new MockStats(5),
                    TaskResourceUsageTrackerType.HEAP_USAGE_TRACKER,
                    new MockStats(0)
                )
            ),
            new SearchShardTaskStats(0, 0, 0, Collections.emptyMap()),
            SearchBackpressureMode.ENFORCED
        );

        SearchBackpressureStats actualStats = service.nodeStats();
        assertEquals(expectedStats, actualStats);
```